tests/Html/FormBuilderTest.php

```diff
@@ -19,40 +33,280 @@ public function setUp() : void
+     */
+    public function testFormInputTextWithLabel()
     {
-        $result = $this->formBuilder->input(type:"text", name:"my-name", value:"my value", options:["id"=>false]);
-        $this->assertEquals('<input id="" name="my-name" type="text" value="my value">', $result);
+        $result = $this->formBuilder->label(name: 'my-input', value: 'my input label');
+        $result = $this->formBuilder->input(type: 'text', name: 'my-input', value: 'my value');
+
+        $this->assertElementIs('input', $result);
+        $this->assertElementAttributeEquals('id', 'my-input', $result);
+        $this->assertElementAttributeEquals('name', 'my-input', $result);
+        $this->assertElementAttributeEquals('type', 'text', $result);
+        $this->assertElementAttributeEquals('value', 'my value', $result);
     }
 
-    public function testInputIdZero()
+    /**
+     * @testdox accepts an empty ID and sets the ID attribute to empty.
+     */
+    public function testFormInputTextIdEmpty()
     {
-        $result = $this->formBuilder->input(type:"text", name:"my-name", value:"my value", options:["id"=>0]);
-        $this->assertEquals('<input id="0" name="my-name" type="text" value="my value">', $result);
+        $result = $this->formBuilder->input(type: 'text', name: 'my-name', value: 'my value', options: ['id' => '']);
+
+        $this->assertElementIs('input', $result);
+        $this->assertElementAttributeEquals('id', '', $result);
+        $this->assertElementAttributeEquals('name', 'my-name', $result);
+        $this->assertElementAttributeEquals('type', 'text', $result);
+        $this->assertElementAttributeEquals('value', 'my value', $result);
     }
 
-    public function testInputIdMissingWithAssociatedLabel()
+    /**
+     * @testdox ignores an ID that is "null".
+     */
+    public function testFormInputTextNull()
     {
-        $result = $this->formBuilder->label(name:"my-input", value:"my input label");
-        $result = $this->formBuilder->input(type:"text", name:"my-input", value:"my value");
-        $this->assertEquals('<input name="my-input" type="text" value="my value" id="my-input">', $result);
+        $result = $this->formBuilder->input(type: 'text', name: 'my-name', value: 'my value', options: ['id' => null]);
+
+        $this->assertElementIs('input', $result);
+        $this->assertElementDoesntHaveAttribute('id', $result);
+        $this->assertElementAttributeEquals('name', 'my-name', $result);
+        $this->assertElementAttributeEquals('type', 'text', $result);
+        $this->assertElementAttributeEquals('value', 'my value', $result);
+    }
+
+    /**
+     * @testdox ignores an ID that is boolean "false".
+     */
+    public function testFormInputTextFalse()
+    {
+        $result = $this->formBuilder->input(type: 'text', name: 'my-name', value: 'my value', options: ['id' => false]);
+
+        $this->assertElementIs('input', $result);
+        $this->assertElementDoesntHaveAttribute('id', $result);
```

Comment:
I'm just saying this changes the previous behavior (which my initial tests were replicating). The idea was to NOT change the original resulting behavior, otherwise hell will break loose... ;)

Reply:
(I know this sounds like me not following my own rules about BC breaks, but they're acceptable if they fix a larger issue IMO :stuck_out_tongue: )

Reply:
I feel the same, but tested this before making the test cases and that was the behevior before the changes I made. So unless your changes with the `if (!empty($name))` broke the behavior at the time, this is going to be a breaking change.

Reply:
Check also https://github.com/wintercms/storm/blob/674733f16629e86cf7f577f29f26fa8c2a0d69c5/src/Html/FormBuilder.php#L693 and https://github.com/wintercms/storm/blob/674733f16629e86cf7f577f29f26fa8c2a0d69c5/src/Html/FormBuilder.php#L695 `return $this->input('submit', null, $value, $options);` - for submit button null value is passed as a name.

Reply:
@bennothommo I'm fine with the BC break. If you're trying to set false or null on an HTML attribute value then you can use the string literal representation of it, so long as we allow for an empty string to be a valid value I'm happy with it.

Reply:
Alright, we're going to treat this as a BC break. Boolean attributes still work as previously defined (ie. adding `['required'] to $options).